server.js

```diff
@@ -157,6 +182,61 @@ server.ext('onPreResponse', function (request, reply) {
     return reply(resp);
 });
 
+// Verifying Same-origin with standard Headers
+// See https://www.owasp.org/index.php?title=Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet&setlang=en
+// Which suggests origin checking in addition to the Double Submit Cookie pattern
+// Exclude the start page which can land
+var same_origin_ignore = [
+    new RegExp('^/$'),
+    new RegExp('^/start$')
+];
+// Register the event handler
+server.ext('onRequest', function (request, reply) {
+    var url = require('url');
+    if (request.headers && !same_origin_ignore.find(path => path.test(request.path))) {
+        // x-forwarded-host should be set by proxies to
+        // preserve the original host where 'host' is
+        // populated with the IP of the proxy. It should be in the
+        // form hostname:port
+        var x_host = request.headers['x-forwarded-host'];
+        var first_xhost = x_host ? x_host.split(",")[0] : undefined;
+        var host = first_xhost || request.headers['host'];
+        var origin = request.headers['origin'] || request.headers['referer'];
+
+        if (!origin) {
+            // OWASP recommends blocking requests for which neither
+            // an origin or a referer is set. However there are a set
+            // of scenarios in which this system; unexpected navigations
+            // start page handlers etc do not set either so we have to ignore
+        } else {
+            if (host) {
+                var p_host = host.split(":");
+                var p_origin = url.parse(origin);
+                if (p_origin.hostname != p_host[0] || p_origin.port != p_host[1]) {
+
+                    var errmsg = 'onRequest[path]: ' + request.path + '\n' +
+                        'onRequest[method]: ' + request.method + '\n' +
+                        'Header[x-forwarded-host]: ' + x_host + '\n' +
+                        'Header[host]: ' + host + '\n' +
+                        'Header[x-forwarded-host]: ' + x_host + '\n' +
+                        'Header[origin]: ' + request.headers['origin'] + '\n' +
+                        'Header[referer]: ' + request.headers['referer'];
+
+                    winston.error('Cross origin request disallowed: ' + p_origin.hostname + ":" + p_origin.port);
```

Comment:
These shouldn't be winston.error calls as this would result in an errbit notification any time a CSRF attack was detected.  Errbit should only be notified when there is an actual problem with the operation of the service, not when someone misuses it.
Probably best dropped to a winston.warn call so that it is at least logged by the frontend.